pkg/kubelet/volumemanager/reconciler/reconstruct_new.go

```diff
@@ -0,0 +1,224 @@
+
+		// Clean up any orphan volumes that failed reconstruction.
+		rc.cleanOrphanVolumes()
+	}
+
+	if !rc.devicePathsReconstructed() {
+		rc.updateReconstructedDevicePaths()
+	}
+}
+
+func (rc *reconciler) readyToUnmount() bool {
+	// Allow unmount only when DSW is fully populated to prevent unmounting volumes that
+	// did not reach DSW yet.
+	if !rc.populatorHasAddedPods() {
+		return false
+	}
+
+	// Allow unmount only when ASW device paths were corrected from node.status to prevent
+	// calling unmount with a wrong devicePath.
+	if !rc.devicePathsReconstructed() {
+		return false
+	}
+	return true
+}
+
+// reconstructVolumes tries to reconstruct the actual state of world by scanning all pods' volume
+// directories from the disk. For the volumes that cannot support or fail reconstruction, it will
+// put the volumes to volumesFailedReconstruction to be cleaned up later when DesiredStateOfWorld
+// is populated.
+func (rc *reconciler) reconstructVolumes() {
+	defer rc.updateLastSyncTime()
+	// Get volumes information by reading the pod's directory
+	podVolumes, err := getVolumesFromPodDir(rc.kubeletPodsDir)
+	if err != nil {
+		klog.ErrorS(err, "Cannot get volumes from disk, skip sync states for volume reconstruction")
+		return
+	}
+	reconstructedVolumes := make(map[v1.UniqueVolumeName]*reconstructedVolume)
+	reconstructedVolumeNames := []v1.UniqueVolumeName{}
+	for _, volume := range podVolumes {
+		if rc.actualStateOfWorld.VolumeExistsWithSpecName(volume.podName, volume.volumeSpecName) {
+			klog.V(4).InfoS("Volume exists in actual state, skip cleaning up mounts", "podName", volume.podName, "volumeSpecName", volume.volumeSpecName)
+			// There is nothing to reconstruct
+			continue
+		}
+		reconstructedVolume, err := rc.reconstructVolume(volume)
+		if err != nil {
+			klog.InfoS("Could not construct volume information", "podName", volume.podName, "volumeSpecName", volume.volumeSpecName, "err", err)
+			// We can't reconstruct the volume. Remember to check DSW after it's fully populated and force unmount the volume when it's orphaned.
+			rc.volumesFailedReconstruction = append(rc.volumesFailedReconstruction, volume)
+			continue
+		}
+		klog.V(4).InfoS("Adding reconstructed volume to actual state and node status", "podName", volume.podName, "volumeSpecName", volume.volumeSpecName)
+		reconstructedVolumeNames = append(reconstructedVolumeNames, reconstructedVolume.volumeName)
+		reconstructedVolumes[reconstructedVolume.volumeName] = reconstructedVolume
+	}
+
+	if len(reconstructedVolumes) > 0 {
+		// Add the volumes to ASW
+		if err = rc.updateStatesNew(reconstructedVolumes); err != nil {
+			klog.ErrorS(err, "Error occurred during reconstruct volume from disk")
+		}
+		// The reconstructed volumes are mounted, hence a previous kubelet must have already put it into node.status.volumesInUse.
+		// Remember to update DSW with this information.
+		rc.desiredStateOfWorld.MarkVolumesReportedInUse(reconstructedVolumeNames)
```

Comment:
we are calling `MarkVolumesReportedInUse` without potentially populating DSOW. Does this have any effect? It seems like the function won't do anything and will be a NO-OP.

Reply:
Good catch! Added a new `rc.volumesNeedReportedInUse` with volumes to mark as in use *after* DSW is populated.